staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go

```diff
@@ -164,76 +203,59 @@ func (c *Controller) sync(name string) error {
+		regenerationCounter.With(map[string]string{"crd": name, "reason": "update"})
 		return nil
 	}
 
-	// update specs of this CRD
-	c.crdSpecs[crd.Name] = newSpecs
+	c.specsByName[crd.Name] = createSpecCache(crd)
 	klog.V(2).Infof("Updating CRD OpenAPI spec because %s changed", name)
-	reason := "add"
-	if updated {
-		reason = "update"
-	}
-	regenerationCounter.With(map[string]string{"crd": name, "reason": reason})
-	return c.updateSpecLocked()
+	regenerationCounter.With(map[string]string{"crd": name, "reason": "add"})
+	c.updateSpecLocked()
+	return nil
 }
 
-func buildVersionSpecs(crd *apiextensionsv1.CustomResourceDefinition, oldSpecs map[string]*spec.Swagger) (map[string]*spec.Swagger, bool, error) {
-	newSpecs := map[string]*spec.Swagger{}
-	anyChanged := false
-	for _, v := range crd.Spec.Versions {
-		if !v.Served {
-			continue
-		}
-		spec, err := builder.BuildOpenAPIV2(crd, v.Name, builder.Options{V2: true})
-		// Defaults must be pruned here for CRDs to cleanly merge with the static
-		// spec that already has defaults pruned
-		spec.Definitions = handler.PruneDefaults(spec.Definitions)
-		if err != nil {
-			return nil, false, err
-		}
-		newSpecs[v.Name] = spec
-		if oldSpecs[v.Name] == nil || !reflect.DeepEqual(oldSpecs[v.Name], spec) {
-			anyChanged = true
-		}
+// updateSpecLocked updates the cached spec graph.
+func (c *Controller) updateSpecLocked() {
+	specList := make([]cached.Data[*spec.Swagger], 0, len(c.specsByName))
+	for crd := range c.specsByName {
+		specList = append(specList, c.specsByName[crd].mergedVersionSpec)
 	}
-	if !anyChanged && len(oldSpecs) == len(newSpecs) {
-		return newSpecs, false, nil
-	}
-
-	return newSpecs, true, nil
-}
 
-// updateSpecLocked aggregates all OpenAPI specs and updates openAPIService.
-// It is not thread-safe. The caller is responsible to hold proper lock (Controller.lock).
-func (c *Controller) updateSpecLocked() error {
-	crdSpecs := []*spec.Swagger{}
-	for _, versionSpecs := range c.crdSpecs {
-		for _, s := range versionSpecs {
-			crdSpecs = append(crdSpecs, s)
+	cache := cached.NewListMerger(func(results []cached.Result[*spec.Swagger]) cached.Result[*spec.Swagger] {
+		localCRDSpec := make([]*spec.Swagger, 0, len(results))
+		for k := range results {
+			if results[k].Err == nil {
+				localCRDSpec = append(localCRDSpec, results[k].Data)
+			}
 		}
-	}
-	mergedSpec, err := builder.MergeSpecs(c.staticSpec, crdSpecs...)
-	if err != nil {
-		return fmt.Errorf("failed to merge specs: %v", err)
-	}
-	return c.openAPIService.UpdateSpec(mergedSpec)
+		mergedSpec, err := builder.MergeSpecs(c.staticSpec, localCRDSpec...)
+		if err != nil {
+			return cached.NewResultErr[*spec.Swagger](fmt.Errorf("failed to merge specs: %v", err))
+		}
+		// A UUID is returned for the etag because we will only
+		// create a new merger when a CRD has changed. A hash based
+		// etag is more expensive because the CRDs are not
+		// premarshalled.
```

Comment:
That's really not great, can't we depend on even just the list of the etags that we've seen sort of like  we do in the aggregator?

Reply:
Not that I'm disagreeing that there may be a better way but isn't it functionally the same as using uuid? 

If none of the downstream etags change, `needsRunning` is false and the cache will not be regenerated. If a downstream etag changes, we regenerate and etag is updated to latest uuid.

The drawback of uuid is if a downstream etag changes and we generate the same merged document, then upstream will see a different etag for the same object and waste computation. I don't think that usually happens though?

Reply:
Yeah, if we have 0 chances of having different etags as an input generate the same output, then the output etag is irrelevant.

Reply:
I'm just not 100% convinced that generation or whatever we use will be good enough at detecting that specific condition (i.e. output will be the exact same).

Reply:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go#L101-L103

Reply:
That's useful but I don't know if that answers my question.
If someone changes [`AdditionalPrinterColumns`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go#L72), do we need to regenerate the whole openapi?

Reply:
There's a tradeoff here. I don't know if we have any other values within a CRD that we could use, but I feel like Generation is the closest approximation for an update to the OpenAPI.

I guess sometimes the Generation may change and the OpenAPI won't, but it's a tradeoff for avoiding to marshal every CRD to fetch a unique etag. The aggregator did not have this problem because its inputs already need to be marshalled to go through an http downloader. To truly ensure that a regeneration will only occur on a unique change, we would need to marshal hash every CRD all time, and compare with the existing hash which seems like quite a bit of computation. The same argument can probably be made for users who like changing the AdditionalPrinterColumn frequently or other parts of the CRD spec not involved in the creation of the OpenAPI. I don't know the answer to this, but I am slightly leaning on the side of using Generation.

Reply:
Thanks, I think we're good.